src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -691,54 +701,86 @@
   (inc! (:updated-catalog performance-metrics))
   (time! (:add-new-catalog performance-metrics)
          (let [catalog-id (:id (add-catalog-metadata! hash catalog received-timestamp))]
-           (update-catalog-associations! catalog-id catalog refs-to-hashes))))
+           (jdbc/update! :certnames {:latest_catalog_id catalog-id} ["id=?" certname-id])
+           (update-catalog-associations! certname-id catalog refs-to-hashes))))
 
 (pls/defn-validated add-catalog!
+  [{:keys [producer_timestamp resources certname] :as catalog} :- catalog-schema
+   received-timestamp :- pls/Timestamp]
+  (time! (:replace-catalog performance-metrics)
+         (jdbc/with-db-transaction []
+           (let [hash (time! (:catalog-hash performance-metrics)
+                             (shash/catalog-similarity-hash catalog))
+                 {certname-id :certname_id
+                  latest-producer-timestamp :producer_timestamp} (latest-catalog-metadata certname)]
+             (inc! (:updated-catalog performance-metrics))
+             (time! (:add-new-catalog performance-metrics)
+                    (let [catalog-id (:id (add-catalog-metadata! hash catalog received-timestamp))]
+
+                      (when-not (some-> latest-producer-timestamp
+                                        (.after (to-timestamp producer_timestamp)))
+                        (let [refs-to-hashes (time! (:resource-hashes performance-metrics)
+                                                    (kitchensink/mapvals shash/resource-identity-hash resources))]
+                          (jdbc/update! :certnames {:latest_catalog_id catalog-id} ["id=?" certname-id])
+                          (update-catalog-associations! certname-id catalog refs-to-hashes)))))
+             hash))))
+
+(pls/defn-validated replace-catalog!
   "Persist the supplied catalog in the database, returning its
    similarity hash."
   ([catalog :- catalog-schema]
-   (add-catalog! catalog (now)))
-  ([{:keys [api_version resources edges certname] :as catalog} :- catalog-schema
+   (replace-catalog! catalog (now)))
+  ([{:keys [producer_timestamp resources certname] :as catalog} :- catalog-schema
     received-timestamp :- pls/Timestamp]
+   (time! (:replace-catalog performance-metrics)
+          (jdbc/with-db-transaction []
+            (let [hash (time! (:catalog-hash performance-metrics)
+                              (shash/catalog-similarity-hash catalog))
+                  {catalog-id :catalog_id
+                   stored-hash :catalog_hash
+                   certname-id :certname_id
+                   latest-producer-timestamp :producer_timestamp} (latest-catalog-metadata certname)]
+              (cond
+                (some-> latest-producer-timestamp
+                        (.after (to-timestamp producer_timestamp)))
+                (log/warnf "Not replacing catalog for certname %s because local data is newer." certname)
+                (= stored-hash hash)
```

Comment:
This shouldn't really happen, should it? Maybe when we change the way we compute the hash?